eslint exhaustive-deps incorrectly labelling constant value as key dependency

### Describe the bug

`@tanstack/eslint-plugin-query@5.60.1`

incorrect exhaustive-deps error for constant values when using them in queryOptions + in a `const` scope

```ts
import { queryOptions } from '@tanstack/react-query';

const EXAMPLE_CONST = 1;

// This works great
export const thisDoesNotShowError = () => ({
  q3: queryOptions({
    queryKey: ['getAll'],
    queryFn: () => Promise.resolve(EXAMPLE_CONST),
  }),
});

// This shows an "The following dependencies are missing in your queryKey: EXAMPLE_CONST" error
export const thisShowsError = {
  q3: queryOptions({
    queryKey: ['getAll'],
    queryFn: () => Promise.resolve(EXAMPLE_CONST),
  }),
};
```

```
npx eslint ./test.ts

/workspace/example.ts
  16:5  error  The following dependencies are missing in your queryKey: EXAMPLE_CONST  @tanstack/query/exhaustive-deps

✖ 1 problem (1 error, 0 warnings)
```

### Your minimal, reproducible example

https://stackblitz.com/edit/vitejs-vite-jx3ix8?file=src%2Ftest.ts

### Steps to reproduce

1. Create an object with properties
2. Use `queryOptions` in one of those property definitions
3. Access a `const` outside the scope of the function
4. ESLint will flag the `@tanstack/query/exhaustive-deps` rule violation


### Expected behavior

Constants should not count as a dependency

### How often does this bug happen?

None

### Screenshots or Videos

_No response_

### Platform

VSCode

### Tanstack Query adapter

None

### TanStack Query version

v5.59.9

### TypeScript version

v5.3.3

### Additional context

Likely related to #5196, which was fixed in #5446

--- a/packages/eslint-plugin-query/src/__tests__/exhaustive-deps.test.ts
+++ b/packages/eslint-plugin-query/src/__tests__/exhaustive-deps.test.ts
@@ -471,8 +471,8 @@ ruleTester.run('exhaustive-deps', rule, {
           queryFn: () => Promise.resolve(EXTERNAL),
         }),
       };  
-      `
-    }
+      `,
+    },
   ],
   invalid: [
     {

--- a/packages/eslint-plugin-query/src/__tests__/exhaustive-deps.test.ts
+++ b/packages/eslint-plugin-query/src/__tests__/exhaustive-deps.test.ts
@@ -460,6 +460,19 @@ ruleTester.run('exhaustive-deps', rule, {
       });
       `,
     },
+    {
+      name: 'should not fail when queryFn inside queryOptions contains a reference to an external variable',
+      code: normalizeIndent`
+      const EXTERNAL = 1;
+
+      export const queries = {
+        foo: queryOptions({
+          queryKey: ['foo'],
+          queryFn: () => Promise.resolve(EXTERNAL),
+        }),
+      };  
+      `
+    }
   ],
   invalid: [
     {
@@ -492,8 +505,10 @@ ruleTester.run('exhaustive-deps', rule, {
     {
       name: 'should fail when no deps are passed (react)',
       code: normalizeIndent`
-        const id = 1;
-        useQuery({ queryKey: ["entity"], queryFn: () => api.getEntity(id) });
+        function Component() {
+          const id = 1;
+          useQuery({ queryKey: ["entity"], queryFn: () => api.getEntity(id) });
+        }
       `,
       errors: [
         {
@@ -504,8 +519,10 @@ ruleTester.run('exhaustive-deps', rule, {
               messageId: 'fixTo',
               data: { result: '["entity", id]' },
               output: normalizeIndent`
-                const id = 1;
-                useQuery({ queryKey: ["entity", id], queryFn: () => api.getEntity(id) });
+                function Component() {
+                  const id = 1;
+                  useQuery({ queryKey: ["entity", id], queryFn: () => api.getEntity(id) });
+                }
               `,
             },
           ],
@@ -515,8 +532,10 @@ ruleTester.run('exhaustive-deps', rule, {
     {
       name: 'should fail when no deps are passed (solid)',
       code: normalizeIndent`
-        const id = 1;
-        createQuery({ queryKey: ["entity"], queryFn: () => api.getEntity(id) });
+        function Component() {
+          const id = 1;
+          createQuery({ queryKey: ["entity"], queryFn: () => api.getEntity(id) });
+        }
       `,
       errors: [
         {
@@ -527,8 +546,10 @@ ruleTester.run('exhaustive-deps', rule, {
               messageId: 'fixTo',
               data: { result: '["entity", id]' },
               output: normalizeIndent`
-                const id = 1;
-                createQuery({ queryKey: ["entity", id], queryFn: () => api.getEntity(id) });
+                function Component() {
+                  const id = 1;
+                  createQuery({ queryKey: ["entity", id], queryFn: () => api.getEntity(id) });
+                }
               `,
             },
           ],
@@ -538,8 +559,10 @@ ruleTester.run('exhaustive-deps', rule, {
     {
       name: 'should fail when deps are passed incorrectly',
       code: normalizeIndent`
-        const id = 1;
-        useQuery({ queryKey: ["entity/\${id}"], queryFn: () => api.getEntity(id) });
+        function Component() {
+          const id = 1;
+          useQuery({ queryKey: ["entity/\${id}"], queryFn: () => api.getEntity(id) });
+        }
       `,
       errors: [
         {
@@ -550,8 +573,10 @@ ruleTester.run('exhaustive-deps', rule, {
               messageId: 'fixTo',
               data: { result: '["entity/${id}", id]' },
               output: normalizeIndent`
-                const id = 1;
-                useQuery({ queryKey: ["entity/\${id}", id], queryFn: () => api.getEntity(id) });
+                function Component() {
+                  const id = 1;
+                  useQuery({ queryKey: ["entity/\${id}", id], queryFn: () => api.getEntity(id) });
+                }
               `,
             },
           ],
@@ -561,9 +586,11 @@ ruleTester.run('exhaustive-deps', rule, {
     {
       name: 'should pass missing dep while key has a template literal',
       code: normalizeIndent`
-        const a = 1;
-        const b = 2;
-        useQuery({ queryKey: [\`entity/\${a}\`], queryFn: () => api.getEntity(a, b) });
+        function Component() {
+          const a = 1;
+          const b = 2;
+          useQuery({ queryKey: [\`entity/\${a}\`], queryFn: () => api.getEntity(a, b) });
+        }
       `,
       errors: [
         {
@@ -574,9 +601,11 @@ ruleTester.run('exhaustive-deps', rule, {
               messageId: 'fixTo',
               data: { result: '[`entity/${a}`, b]' },
               output: normalizeIndent`
-                const a = 1;
-                const b = 2;
-                useQuery({ queryKey: [\`entity/\${a}\`, b], queryFn: () => api.getEntity(a, b) });
+                function Component() {
+                  const a = 1;
+                  const b = 2;
+                  useQuery({ queryKey: [\`entity/\${a}\`, b], queryFn: () => api.getEntity(a, b) });
+                }
               `,
             },
           ],
@@ -586,14 +615,16 @@ ruleTester.run('exhaustive-deps', rule, {
     {
       name: 'should fail when dep exists inside setter and missing in queryKey',
       code: normalizeIndent`
-        const [id] = React.useState(1);
-        useQuery({
+        function Component() {
+          const [id] = React.useState(1);
+          useQuery({
             queryKey: ["entity"],
             queryFn: () => {
-                const { data } = axios.get(\`.../\${id}\`);
-                return data;
+              const { data } = axios.get(\`.../\${id}\`);
+              return data;
             }
-        });
+          });
+        }
       `,
       errors: [
         {
@@ -604,14 +635,16 @@ ruleTester.run('exhaustive-deps', rule, {
               messageId: 'fixTo',
               data: { result: '["entity", id]' },
               output: normalizeIndent`
-                const [id] = React.useState(1);
-                useQuery({
+                function Component() {
+                  const [id] = React.useState(1);
+                  useQuery({
                     queryKey: ["entity", id],
                     queryFn: () => {
-                        const { data } = axios.get(\`.../\${id}\`);
-                        return data;
+                      const { data } = axios.get(\`.../\${id}\`);
+                      return data;
                     }
-                });
+                  });
+                }
               `,
             },
           ],
@@ -713,9 +746,11 @@ ruleTester.run('exhaustive-deps', rule, {
     {
       name: 'should fail when a queryKey is a reference of an array expression with a missing dep',
       code: normalizeIndent`
-        const x = 5;
-        const queryKey = ['foo']
-        useQuery({ queryKey, queryFn: () => x })
+        function Component() {
+          const x = 5;
+          const queryKey = ['foo']
+          useQuery({ queryKey, queryFn: () => x })
+        }
       `,
       errors: [
         {
@@ -728,9 +763,11 @@ ruleTester.run('exhaustive-deps', rule, {
                 result: "['foo', x]",
               },
               output: normalizeIndent`
-                const x = 5;
-                const queryKey = ['foo', x]
-                useQuery({ queryKey, queryFn: () => x })
+                function Component() {
+                  const x = 5;
+                  const queryKey = ['foo', x]
+                  useQuery({ queryKey, queryFn: () => x })
+                }
               `,
             },
           ],
@@ -758,12 +795,14 @@ ruleTester.run('exhaustive-deps', rule, {
     {
       name: 'should fail if queryFn is using multiple object props when only one of them is in the queryKey',
       code: normalizeIndent`
-        const state = { foo: 'foo', bar: 'bar' }
+        function Component() {
+          const state = { foo: 'foo', bar: 'bar' }
 
-        useQuery({
+          useQuery({
             queryKey: ['state', state.foo],
             queryFn: () => Promise.resolve({ foo: state.foo, bar: state.bar })
-        })
+          })
+        }
       `,
       errors: [
         {
@@ -771,12 +810,14 @@ ruleTester.run('exhaustive-deps', rule, {
             {
               messageId: 'fixTo',
               output: normalizeIndent`
-              const state = { foo: 'foo', bar: 'bar' }
+              function Component() {
+                const state = { foo: 'foo', bar: 'bar' }
 
-              useQuery({
+                useQuery({
                   queryKey: ['state', state.foo, state.bar],
                   queryFn: () => Promise.resolve({ foo: state.foo, bar: state.bar })
-              })
+                })
+              }
             `,
             },
           ],
@@ -788,14 +829,16 @@ ruleTester.run('exhaustive-deps', rule, {
     {
       name: 'should fail if queryFn is invalid while using FunctionExpression syntax',
       code: normalizeIndent`
-        const id = 1;
+        function Component() {
+          const id = 1;
 
-        useQuery({
+          useQuery({
             queryKey: [],
             queryFn() {
               Promise.resolve(id)
             }
-        })
+          });
+        }
       `,
       errors: [
         {
@@ -803,14 +846,16 @@ ruleTester.run('exhaustive-deps', rule, {
             {
               messageId: 'fixTo',
               output: normalizeIndent`
-                const id = 1;
+                function Component() {
+                  const id = 1;
 
-                useQuery({
+                  useQuery({
                     queryKey: [id],
                     queryFn() {
                       Promise.resolve(id)
                     }
-                })
+                  });
+                }
               `,
             },
           ],

--- a/packages/eslint-plugin-query/src/rules/exhaustive-deps/exhaustive-deps.utils.ts
+++ b/packages/eslint-plugin-query/src/rules/exhaustive-deps/exhaustive-deps.utils.ts
@@ -13,7 +13,7 @@ export const ExhaustiveDepsUtils = {
     const component = ASTUtils.getFunctionAncestor(sourceCode, node)
 
     if (component === undefined) {
-        return false
+      return false
     }
 
     if (

--- a/packages/eslint-plugin-query/src/rules/exhaustive-deps/exhaustive-deps.utils.ts
+++ b/packages/eslint-plugin-query/src/rules/exhaustive-deps/exhaustive-deps.utils.ts
@@ -12,8 +12,11 @@ export const ExhaustiveDepsUtils = {
     const { sourceCode, reference, scopeManager, node } = params
     const component = ASTUtils.getFunctionAncestor(sourceCode, node)
 
+    if (component === undefined) {
+        return false
+    }
+
     if (
-      component !== undefined &&
       !ASTUtils.isDeclaredInNode({
         scopeManager,
         reference,

--- a/packages/eslint-plugin-query/src/utils/ast-utils.ts
+++ b/packages/eslint-plugin-query/src/utils/ast-utils.ts
@@ -241,7 +241,13 @@ export const ASTUtils = {
     node: TSESTree.Node,
   ) {
     for (const ancestor of sourceCode.getAncestors(node)) {
-      if (ancestor.type === AST_NODE_TYPES.FunctionDeclaration) {
+      if (
+        ASTUtils.isNodeOfOneOf(ancestor, [
+          AST_NODE_TYPES.FunctionDeclaration,
+          AST_NODE_TYPES.FunctionExpression,
+          AST_NODE_TYPES.ArrowFunctionExpression,
+        ])
+      ) {
         return ancestor
       }
 

